src/ovs/aca_ovs_programmer.cpp

```diff
@@ -34,51 +39,78 @@ ACA_OVS_Programmer &ACA_OVS_Programmer::get_instance()
   return instance;
 }
 
-int ACA_OVS_Programmer::setup_bridges()
+int ACA_OVS_Programmer::setup_ovs_bridges_if_need()
 {
-  ACA_LOG_DEBUG("ACA_OVS_Programmer::setup_bridges ---> Entering\n");
+  ACA_LOG_DEBUG("ACA_OVS_Programmer::setup_ovs_bridges_if_need ---> Entering\n");
 
   ulong not_care_culminative_time;
   int overall_rc = EXIT_SUCCESS;
 
-  // TODO: confirm OVS is up and running, if not, try to start it
+  // -----critical section starts-----
+  // (exclusive access to br-int and br-tun creation status signaled by locking setup_ovs_bridges_mutex):
+  setup_ovs_bridges_mutex.lock();
+
+  // check to see if br-int and br-tun is already there
+  execute_ovsdb_command("br-exists br-int", not_care_culminative_time, overall_rc);
+  bool br_int_existed = (overall_rc == EXIT_SUCCESS);
+
+  execute_ovsdb_command("br-exists br-tun", not_care_culminative_time, overall_rc);
+  bool br_tun_existed = (overall_rc == EXIT_SUCCESS);
+
+  overall_rc = EXIT_SUCCESS;
+
+  // case 1: both br-int and br-tun exist
+  if (br_int_existed && br_int_existed) {
+    // nothing to do
+  }
+
+  // case 2: both br-int and br-tun not there, create them
+  else if (!br_int_existed && !br_int_existed) {
+    // create br-int and br-tun bridges
+    execute_ovsdb_command("add-br br-int", not_care_culminative_time, overall_rc);
 
-  // TODO: see if br-int and br-tun is already there, one option is the delete
-  // them and start everything from scratch
+    execute_ovsdb_command("add-br br-tun", not_care_culminative_time, overall_rc);
 
-  // create br-int and br-tun bridges
-  execute_ovsdb_command("add-br br-int", not_care_culminative_time, overall_rc);
+    // execute_openflow_command("del-flows br-tun", not_care_culminative_time, overall_rc);
 
-  execute_ovsdb_command("add-br br-tun", not_care_culminative_time, overall_rc);
+    // create and connect the patch ports between br-int and br-tun
+    execute_ovsdb_command("add-port br-int patch-tun", not_care_culminative_time, overall_rc);
 
-  execute_openflow_command("del-flows br-tun", not_care_culminative_time, overall_rc);
+    execute_ovsdb_command("set interface patch-tun type=patch",
+                          not_care_culminative_time, overall_rc);
 
-  // create and connect the patch ports between br-int and br-tun
-  execute_ovsdb_command("add-port br-int patch-tun", not_care_culminative_time, overall_rc);
+    execute_ovsdb_command("set interface patch-tun options:peer=patch-int",
```

Comment:
At this time, patch-int doesn't exist yet. Could it be a problem?

Reply:
it will be okay :)
ovs will return an error for this call as expected.
the next call will add the patch-int port and the problem will fix itself.

Reply:
I see. So the very first call will always generate some false negative error message. Could we make it self-contained by retrying until it succeeds, not a blocking issue, just something we could consider.

Reply:
I change some ordering and it doesn't help :(.

To clarify (I was a little off on which call would generate error):

ovs-vsctl add-br br-int
ovs-vsctl add-br br-tun
ovs-vsctl add-port br-int patch-tun <- error: could not open network device patch-tun (No such device)
ovs-vsctl set interface patch-tun type=patch <- would fix the above, swap the order to call it first doesn't work
ovs-vsctl set interface patch-tun options:peer=patch-int <- no error